astropy/io/fits/header.py

```diff
@@ -1537,9 +1525,28 @@ def remove(self, keyword):
         keyword : str
             The keyword of which to remove the first instance in the header
 
+        ignore_missing : bool, optional
+            When True, ignores missing keywords.  Otherwise, if the keyword is not present
+            in the header a KeyError is raised.
+
+        all : bool, optional
+            When True, all instances of keyword will be removed.
+            Otherwise only the first instance of the given keyword is removed.
+
         """
+        keyword = Card.normalize_keyword(keyword)
+        try:
+            if keyword not in self._keyword_indices:
+                raise KeyError("Keyword '%s' not in the header." % keyword)
+
+            del self[self._keyword_indices[keyword][0]]
+            if all:
+                while keyword in self._keyword_indices:
+                    del self[self._keyword_indices[keyword][0]]
+        except KeyError:
+            if not ignore_missing:
+                raise
```

Comment:
I see now, I think I misled you by forgetting, myself, that keyword_indices is a defaultdict, and won't raise a `KeyError`.

In this case adding the try/except overcomplicates things.  Just change this to:

```python
if keyword in self._keyword_indices:
    del ...
    if all:
        ...etc...
else:
    if not ignore_missing:
        raise KeyError(....)
```

Make the text for that last `KeyError` the same as [this one](https://github.com/astropy/astropy/pull/4445/files#diff-2a28d7f27c4e8283e7899622e63431c4R225).  You can see that the end result is the exact same logic, but vastly simplified.